[16.0] refactor quotations rest api

nicolas-delbovier-acsone opened the pull request:
This PR introduces several refactorings to improve module isolation and extend the Shopinvader API:

1. **`sale_typology` module creation:** Isolates the sale order typology logic.
2. **`sale_quotation` extension:** Updates `sale_quotation` to depend on the new `sale_typology` module.
3. **`sale_cart_quotation` module creation:** Decouples `sale_quotation` from `sale_cart`.
4. **`shopinvader_api_quotation` enhancement:** Enables line creation, update, and deletion within the quotation API.
5. **`shopinvader_api_cart_quotation` module extraction:** Moves the cart-to-quotation transformation logic from `shopinvader_api_quotation` to its own dedicated module.

nicolas-delbovier-acsone committed:
[REF] sale_quotation: refactor the quotation mechanism.

* Extracts the shop_order_mode of products in a new module
* Enables to choose a combined shop order mode (quotation or direct sale)
* Adds a migration script to translate the products and templates order mode infos from the old to the new fields

nicolas-delbovier-acsone committed:
[REF] sale_quotation: refactor the quotation mechanism.

* Extracts the shop_order_mode of products in a new module
* Enables to choose a combined shop order mode (quotation or direct sale)
* Adds a migration script to translate the products and templates order mode infos from the old to the new fields

nicolas-delbovier-acsone committed:
[ADD] sale_typology: extract the concept of "typology" in sale.order in a separate module

nicolas-delbovier-acsone committed:
[ADD] sale_typology: extract the concept of "typology" in sale.order in a separate module

nicolas-delbovier-acsone committed:
[IMP] sale_quotation: use the "typology" field to delcare the sale order as a quotation

nicolas-delbovier-acsone committed:
[IMP] sale_quotation: use the "typology" field to delcare the sale order as a quotation

nicolas-delbovier-acsone committed:
[ADD] sale_cart_quotation: introduce module to decouple sale_quotation and sale_cart

nicolas-delbovier-acsone committed:
[ADD] sale_cart_quotation: introduce module to decouple sale_quotation and sale_cart

nicolas-delbovier-acsone committed:
[IMP] shopinvader_api_quotation: remove sale_cart dependency and enhance API

- put shopinvader_sale_state as a dependency in order to remove sale_cart because we need the field sale.order.shopinvader_state

This commit introduces significant enhancements to the REST API for quotation management, including:

- **Create Method:** Added a new endpoint to create quotations.
- **Line Management Methods:** Implemented dedicated endpoints for adding, deleting, or updating single or multiple lines within a quotation.
- **Update Extension:** Extended the existing update method to support the addition of new lines

nicolas-delbovier-acsone committed:
[IMP] shopinvader_api_quotation: remove sale_cart dependency and enhance API

- put shopinvader_sale_state as a dependency in order to remove sale_cart because we need the field sale.order.shopinvader_state

This commit introduces significant enhancements to the REST API for quotation management, including:

- **Create Method:** Added a new endpoint to create quotations.
- **Line Management Methods:** Implemented dedicated endpoints for adding, deleting, or updating single or multiple lines within a quotation.
- **Update Extension:** Extended the existing update method to support the addition of new lines

nicolas-delbovier-acsone committed:
[ADD] shopinvader_api_cart_quotation: add a module to isolate the logic of transforming a cart to a quotation.

nicolas-delbovier-acsone committed:
[ADD] shopinvader_api_cart_quotation: add a module to isolate the logic of transforming a cart to a quotation.

lmignon commented on sale_cart/models/sale_order.py:
```diff
@@ -8,7 +8,7 @@ class SaleOrder(models.Model):
     _inherit = "sale.order"
 
-    typology = fields.Selection([("sale", "Sale"), ("cart", "Cart")], default="sale")
+    typology = fields.Selection(selection_add=[("cart", "Cart")])
```
@nicolas-delbovier-acsone The ondelete stragegy is missing here

lmignon commented on sale_cart_quotation/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1 @@
+Glue module between sale_quotation and sale_cart
```
This module is not a glue module only. It adds the functionality of requesting a quotation for a cart. This should be specified into the description.

lmignon commented on sale_cart_quotation/views/sale_order.xml:
```diff
@@ -0,0 +1,12 @@
+<!-- Copyright 2025 ACSONE SA/NV
+     License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). -->
+<odoo>
+    <record id="sale.action_quotations" model="ir.actions.act_window">
```
IMO a button 'Request Quotation' should be added to call the `action_request_quotation` method. It should be visible only if the so typology is 'cart'. In this way the logic is also available from the back-office and not only used by the shopinvader api.

lmignon commented on sale_quotation/models/sale_order.py:
```diff
@@ -24,7 +25,8 @@ class SaleOrder(models.Model):
     )
-    shop_only_quotation = fields.Boolean(compute="_compute_shop_only_quotation")
+
+    typology = fields.Selection(selection_add=[("quote", "Quote")], default="quote")
```
The ondelete strategy is missing

lmignon commented on shopinvader_api_quotation/__manifest__.py:
```diff
@@ -13,7 +13,7 @@
         "shopinvader_api_cart",
         "shopinvader_api_sale",
-        "sale_cart",
+        "shopinvader_sale_state",
```
Why do you need this dependency?

nicolas-delbovier-acsone replied:
Because of the overwrite of the `shopinvader_state` field inside models/sale_order.py
```
class SaleOrder(models.Model):
    _inherit = "sale.order"

    shopinvader_state = fields.Selection(
        selection_add=[
            ("estimating", "Estimating Quotation"),
            ("estimated", "Estimated Quotation"),
        ],
        ondelete={
            "estimating": "cascade",
            "estimated": "cascade",
        },
    )

    def _compute_shopinvader_state_depends(self):
        return super()._compute_shopinvader_state_depends() + ("quotation_state",)

    def _get_shopinvader_state(self):
        self.ensure_one()
        if self.quotation_state == "customer_request":
            return "estimating"
        if self.quotation_state == "waiting_acceptation":
            return "estimated"
        return super()._get_shopinvader_state()

```
The module would not install without this dependency: `AssertionError: Field sale.order.shopinvader_state without selection`

lmignon commented on shopinvader_api_quotation/dependencies.py:
```diff
@@ -0,0 +1,140 @@
+                for line in vals.pop("lines")
+            ]
+
+        quotation = self.env["sale.order"].create(vals)
```
IMO we should force the `typology` to quote to ensure that the created SO is really a quotation. Imagine the case where the default for the typology field is changed by another addon....

lmignon commented on shopinvader_api_quotation/dependencies.py:
```diff
@@ -0,0 +1,140 @@
+
+    def _update(self, quotation_id, data):
+        quotation = self._get(quotation_id)
+        quotation.write(data.to_sale_order_vals())
```
If the lines fields is provided into the update method, we should consider that at the end of the process, our quotation should only contains the same number of lines as the one provided...
* If the line has an id -> update existing line (raise if not found)
* If the line has no id -> create a new one
* All existing lines not found in lines should be removed

To know which fields are part of the initial json you can use the method `values = self.model_dump(exclude_unset=True)` values is a dict with only fields provided into the json.

lmignon commented on shopinvader_api_quotation/schemas/sale.py:
```diff
@@ -13,12 +17,63 @@ def from_sale_order(cls, odoo_rec):
+    name: str
+    lines: list[QuotationLineCreateRequest] | None = None
     client_order_ref: str | None = None
+    typology: str
```
typology should not be part of the create request.... It can't be manipulated by the api

nicolas-delbovier-acsone committed:
[IMP] sale_cart: add ondelete strategy on typology field

nicolas-delbovier-acsone committed:
[IMP] sale_cart: add ondelete strategy on typology field

nicolas-delbovier-acsone committed:
[IMP] sale_cart_quotation: add a button to trigger the "request quotation" action in the form view of sale.order + refine module description

nicolas-delbovier-acsone committed:
[IMP] sale_cart_quotation: add a button to trigger the "request quotation" action in the form view of sale.order + refine module description

nicolas-delbovier-acsone committed:
[IMP] sale_quotation: add ondelete strategy to typology field in sale.order.

nicolas-delbovier-acsone committed:
[IMP] sale_quotation: add ondelete strategy to typology field in sale.order.

nicolas-delbovier-acsone committed:
[IMP] shopinvader_api_quotation: improve update logic + small fixes.

- Remove typology field from quotation create request
- Implement following logic in the update process:
  -If the line has an id -> update existing line (raise if not found)
  -If the line has no id -> create a new one
  -All existing lines not found in lines should be removed
- Small refactoring to avoid code dup
- Test invalid line ids in unit tests

nicolas-delbovier-acsone committed:
[IMP] shopinvader_api_quotation: improve update logic + small fixes.

- Remove typology field from quotation create request
- Implement following logic in the update process:
  -If the line has an id -> update existing line (raise if not found)
  -If the line has no id -> create a new one
  -All existing lines not found in lines should be removed
- Small refactoring to avoid code dup
- Test invalid line ids in unit tests

nicolas-delbovier-acsone committed:
[IMP] shopinvader_product_order_mode: use field dependency instead of hacking the write.

nicolas-delbovier-acsone committed:
[IMP] shopinvader_product_order_mode: use field dependency instead of hacking the write.

nicolas-delbovier-acsone committed:
[FIX] fix typo 'unabled' -> 'enabled'

nicolas-delbovier-acsone committed:
[FIX] fix typo 'unabled' -> 'enabled'

nicolas-delbovier-acsone committed:
[IMP] fix typo 'unabled' -> 'enabled'

nicolas-delbovier-acsone committed:
[IMP] fix typo 'unabled' -> 'enabled'

nicolas-delbovier-acsone commented:
The last force push aimed at rebasing this PR to #1594.

nicolas-delbovier-acsone committed:
[FIX] sale_quotation: Remove redundant unit test

Deletes an obsolete unit test that was migrated to the `sale_cart_quotation` module.

nicolas-delbovier-acsone committed:
[FIX] sale_quotation: Remove redundant unit test

Deletes an obsolete unit test that was migrated to the `sale_cart_quotation` module.

nicolas-delbovier-acsone committed:
[FIX] sale_quotation: Restore standard SO confirmation; introduce optional custom workflow

This commit addresses a critical regression where the standard Odoo sales order confirmation flow was unintentionally broken, making it impossible to directly confirm a newly created SO.

The previous logic implicitly forced a custom quotation process even when not desired. This refactor introduces an explicit opt-in mechanism for the new quotation workflow.

Key changes:
- **Default Typology:** Removed a redundant test in `sale_typology` as the `typology` field's default value is now correctly handled and overridden by `sale_quotation`.
- **Workflow Toggle:** Added a `use_customer_quotation_workflow` boolean field to `sale.order`, allowing users to enable or disable the custom quotation stages.
- **Controlled Confirmation:** When `use_customer_quotation_workflow` is active, a UI wizard now prompts the user to confirm their intent if they attempt to confirm a quotation not yet in the 'waiting_acceptation' state. This preserves the standard Odoo workflow for non-workflow-enabled orders while providing a gentle warning for custom flows.

nicolas-delbovier-acsone committed:
[FIX] sale_quotation: Restore standard SO confirmation; introduce optional custom workflow

This commit addresses a critical regression where the standard Odoo sales order confirmation flow was unintentionally broken, making it impossible to directly confirm a newly created SO.

The previous logic implicitly forced a custom quotation process even when not desired. This refactor introduces an explicit opt-in mechanism for the new quotation workflow.

Key changes:
- **Default Typology:** Removed a redundant test in `sale_typology` as the `typology` field's default value is now correctly handled and overridden by `sale_quotation`.
- **Workflow Toggle:** Added a `use_customer_quotation_workflow` boolean field to `sale.order`, allowing users to enable or disable the custom quotation stages.
- **Controlled Confirmation:** When `use_customer_quotation_workflow` is active, a UI wizard now prompts the user to confirm their intent if they attempt to confirm a quotation not yet in the 'waiting_acceptation' state. This preserves the standard Odoo workflow for non-workflow-enabled orders while providing a gentle warning for custom flows.

nicolas-delbovier-acsone committed:
[IMP] sale_quotation: improve UI of the sale_order form.

* add a button to accept the quotation
* add some margin to the workflow text
* Show quotation_state in the title instead of messing with the status bar

nicolas-delbovier-acsone committed:
[IMP] sale_quotation: improve UI of the sale_order form.

* add a button to accept the quotation
* add some margin to the workflow text
* Show quotation_state in the title instead of messing with the status bar

nicolas-delbovier-acsone committed:
[IMP] sale_quotation: rename method for clarity.

Changes `action_confirm_quotation` to `action_accept_quotation`.

nicolas-delbovier-acsone committed:
[IMP] sale_quotation: rename method for clarity.

Changes `action_confirm_quotation` to `action_accept_quotation`.

lmignon committed:
[IMP] sale_cart: Makes menu more flexible

Normal quotations are sale.order with an other typology than 'cart'. This allows the use of others values for typology than 'cart' and 'sale'.

lmignon committed:
[IMP] sale_cart: Makes menu more flexible

Normal quotations are sale.order with an other typology than 'cart'. This allows the use of others values for typology than 'cart' and 'sale'.

lmignon committed:
[IMP]sale_quotation: Improves menu.

Ensures menu are properly working if sale_cart or sale_quoration are installed but not the orher one

lmignon committed:
[IMP]sale_quotation: Improves menu.

Ensures menu are properly working if sale_cart or sale_quoration are installed but not the orher one

lmignon committed:
[IMP] sale_quotation: Improves UI/UX

Use a switch button to activate / deactivate the user workflow for quotation. All the logic is now driven by the field and we no more need to manipulate the typology

lmignon committed:
[IMP] sale_quotation: Improves UI/UX

Use a switch button to activate / deactivate the user workflow for quotation. All the logic is now driven by the field and we no more need to manipulate the typology

lmignon committed:
[IMP] sale_quotation: Defines user's specific actions

Add methods triggered by the cutomer to manage the requested quotation. Ensures that methods called by customer are allowed according the the quotation state

lmignon committed:
[IMP] sale_quotation: Defines user's specific actions

Add methods triggered by the cutomer to manage the requested quotation. Ensures that methods called by customer are allowed according the the quotation state

lmignon committed:
[IMP] sale_cart_quotation: Uses new actions declared for customer use only

lmignon committed:
[IMP] sale_cart_quotation: Uses new actions declared for customer use only

lmignon committed:
[IMP] shopinvader_api_cart_quotation: CUses new actions declared for customer use only

lmignon committed:
[IMP] shopinvader_api_cart_quotation: CUses new actions declared for customer use only

lmignon committed:
[IMP] shopinvader_api_quotation: Improves API

Adds customer methods used to manage the quotation flow.
Ensures proper http status is returned in case where the called method is not allowed according to the quotation state

lmignon committed:
[IMP] shopinvader_api_quotation: Improves API

Adds customer methods used to manage the quotation flow.
Ensures proper http status is returned in case where the called method is not allowed according to the quotation state

lmignon committed:
[IMP] sale_quotation, shopinvader_api_quotation: add FR translations

lmignon committed:
[IMP] sale_quotation, shopinvader_api_quotation: add FR translations

lmignon committed:
[IMP] sale_quotation: Improves UI

lmignon committed:
[IMP] sale_quotation: Improves UI

lmignon committed:
[IMP] sale_quotation: Improves UI

lmignon committed:
[IMP] sale_quotation: Improves UI

lmignon committed:
[IMP] sale_quotation: Translations

lmignon committed:
[IMP] sale_quotation: Translations

lmignon committed:
[IMP] shopinvader_api_quotation: No download while the quotation is not sent to the customer

lmignon committed:
[IMP] shopinvader_api_quotation: No download while the quotation is not sent to the customer

lmignon committed:
[IMP] shopinvader_api_quotation: Quotation name is not editable.

The customer can only fill the client ref as rederence. The name is reserved for Odoo

lmignon committed:
[IMP] shopinvader_api_quotation: Quotation name is not editable.

The customer can only fill the client ref as rederence. The name is reserved for Odoo

lmignon committed:
[IMP]shopinvader_api_quotation: Refactor code to split sale and quotation schemas

lmignon committed:
[IMP]shopinvader_api_quotation: Refactor code to split sale and quotation schemas

lmignon committed:
[IMP] shopinvader_api_quotation: Add search params

Allows to search on quotation_state and client_order_ref

lmignon committed:
[IMP] shopinvader_api_quotation: Add search params

Allows to search on quotation_state and client_order_ref

lmignon committed:
[IMP] shopinvader_api_quotation: Add the quotation_state into the sale info

lmignon committed:
[IMP] shopinvader_api_quotation: Add the quotation_state into the sale info

lmignon committed:
[IMP] shopinvader_api_quotation: Implement partial update and add note

lmignon committed:
[IMP] shopinvader_api_quotation: Implement partial update and add note

lmignon committed:
[FIX] sale_quotation, shopinvader_api_quotation: Add pot file

complete pot files are required to get the translations working. Indeed, when a transation file is loaded by the polib, the entry is marked as obsolete if it's not present into the pot file. obsolete entries are not loaded by odoo

lmignon committed:
[FIX] sale_quotation, shopinvader_api_quotation: Add pot file

complete pot files are required to get the translations working. Indeed, when a transation file is loaded by the polib, the entry is marked as obsolete if it's not present into the pot file. obsolete entries are not loaded by odoo

paradoxxxzero commented on sale_quotation/models/sale_order.py:
```diff
@@ -42,29 +84,164 @@ def _compute_quotation_state(self):
+            customer_quotations = self.filtered("use_customer_quotation_workflow")
+            if customer_quotations:
+                self.action_customer_reset_quotation_to_draft()
+        return super(SaleOrder, other_quotations).action_draft()
```
It seems that `self` is not removed from `other_quotations` resulting in a double call to `action_draft` in case of `bypass_customer_quotation=False`.

lmignon replied:
good catch!

paradoxxxzero commented on sale_quotation/models/sale_order.py:
```diff
@@ -42,29 +84,164 @@ def _compute_quotation_state(self):
+                }
+            else:
+                customer_quotations.action_customer_accept_quotation()
+        return super(SaleOrder, other_quotations).action_confirm()
```
It seems that `self` is not removed from `other_quotations` resulting in a double call to `action_confirm` in case of `action_customer_accept_quotation()` call.

paradoxxxzero commented on sale_quotation/models/sale_order.py:
```diff
@@ -42,29 +84,164 @@ def _compute_quotation_state(self):
+        if not self.env.context.get("bypass_customer_quotation", False):
+            customer_quotations = self.filtered("use_customer_quotation_workflow")
+            if customer_quotations:
+                self.action_customer_reset_quotation_to_draft()
```
Shouldn't this be:

```suggestion
                customer_quotations.action_customer_reset_quotation_to_draft()
```

?

paradoxxxzero commented on sale_quotation/models/sale_order.py:
```diff
@@ -42,29 +84,164 @@ def _compute_quotation_state(self):
+    def action_quotation_sent(self):
+        customer_quotations = self.filtered("use_customer_quotation_workflow")
+        if customer_quotations:
+            self._check_customer_action_allowed("quotation_sent")
```
```suggestion
            customer_quotations._check_customer_action_allowed("quotation_sent")
```

lmignon committed:
[FIX] sale_quotation: Avoid double call to actions

lmignon committed:
[FIX] sale_quotation: Avoid double call to actions

lmignon commented:
@paradoxxxzero fixed

sebastienbeau commented on shopinvader_api_quotation/dependencies.py:
```diff
@@ -0,0 +1,284 @@
+)
+
+
+def InvalidQuotationStateErrorWrapper(func):
```
Interesting code. Creating a specific error so we can catch it and return a clean http error code is smart.

In long term, maybe we should be able to extend the "convert_exception_to_status_body"
https://github.com/OCA/rest-framework/blob/34c06aab48512b6d7bda8ec797c45ae219a2fc59/fastapi/error_handlers.py#L20 so we can just extend this and it will raise the right http error without needed a wrapper.

The issue with the wrapper is when an other endpoint do an action and you didn't have set the decoration (because you didn't expect this case) you will not have the right http error.

Note a blocking point, it's just a comment to share the idea. If you think it's an interesting one we can to a PR in fastapi to implement it.

sebastienbeau commented on sale_quotation/models/sale_order.py:
```diff
@@ -42,29 +84,166 @@ def _compute_quotation_state(self):
+                customer_quotations.action_customer_accept_quotation()
+        return super(SaleOrder, other_quotations).action_confirm()
+
+    def action_draft(self):
+        other_quotations = self
+        if not self.env.context.get("bypass_customer_quotation", False):
+            customer_quotations = self.filtered("use_customer_quotation_workflow")
+            other_quotations = self - customer_quotations
+            if customer_quotations:
+                customer_quotations.action_customer_reset_quotation_to_draft()
+        return super(SaleOrder, other_quotations).action_draft()
```
I always try to avoid function that call themself in both way (it's harder to read, and it can generate unexpected random issue)

For example if you have a custom module that do the following

```
def action_draft(self):
   for sale in self:
       sale._action_draft_count += 1
   super().action_draft()
```

Then depending on the order of loading the module, the increment code can be called before or after the inherit in sale_quotation.

In case that it's called before, if on a quotation you click on action_draft, it will increment a first time, then it will call "action_customer_reset_quotation_to_draft" that will call "action_draft" and increment a second time.

The crazy thing is that if the module are loaded in a different order (and the increment is call with the super), you will not have the issue, so have random bug (True bad experience).


Maybe we can simplify this logic and avoid the extra code of "bypass_customer_quotation"

Something like:

```suggestion

    def action_customer_reset_quotation_to_draft(self):
       self._check_customer_action_allowed("reset_quotation_to_draft")
       return self.action_draft()

    def action_draft(self):
        customer_quotations = self.filtered("use_customer_quotation_workflow")
        customer_quotations.quotation_state = "draft"
        customer_quotations.filtered(lambda so: so.typology != "quote").typology = "quote"
        return super().action_draft()
```

What do you think ?

lmignon replied:
With your approach, the logic of resetting a customer quotation to draft is no more isolated into a dedicated method....  The current implementation also ensure that everything stay consistent if you click on the 'main' button for action_draft or on the customer one....

sebastienbeau replied:
Maybe we should have something like 


```
    def _set_customer_reset_quotation_to_draft(self):
        self.quotation_state = "draft"
        self.filtered(lambda so: so.typology != "quote").typology = "quote"

    def action_customer_reset_quotation_to_draft(self):
        self._check_customer_action_allowed("reset_quotation_to_draft")
        self._set_customer_reset_quotation_to_draft()
        return self.with_context(skip_set_customer_reset_quotation_to_draft=True).action_draft()

    def action_draft(self):
        if not self._context.get("skip_set_customer_reset_quotation_to_draft"):
            self.filtered("use_customer_quotation_workflow")._set_customer_reset_quotation_to_draft()
        return super().action_draft()
```

Or



```
    def _set_customer_reset_quotation_to_draft(self):
        self.quotation_state = "draft"
        self.filtered(lambda so: so.typology != "quote").typology = "quote"

    def action_customer_reset_quotation_to_draft(self):
        self._check_customer_action_allowed("reset_quotation_to_draft")
        return self.action_draft()

    def action_draft(self):
        self.filtered("use_customer_quotation_workflow")._set_customer_reset_quotation_to_draft()
        return super().action_draft()
```


What do you think ?

sebastienbeau commented on sale_quotation/models/sale_order.py:
```diff
@@ -42,29 +84,166 @@ def _compute_quotation_state(self):
+    def action_customer_reset_quotation_to_draft(self):
+        self._check_customer_action_allowed("reset_to_draft")
+        self.quotation_state = "draft"
+        self.state = "draft"
```
setting draft here is useless as action_draft will do it

```suggestion
```

sebastienbeau commented on sale_quotation/models/sale_order.py:
```diff
@@ -42,29 +84,166 @@ def _compute_quotation_state(self):
+        self._check_customer_action_allowed("accept_quotation")
         self.quotation_state = "accepted"
+        self.typology = "sale"
+        return self.with_context(
+            bypass_customer_quotation=True,
+        ).action_confirm()
```
This change with the previous behavior.
Indeed before this change an accepted quotation was staying in draft (as maybe we wait for extra action like the payment)

The two customer we have, for now make this difference, they want to know that the customer have accepted the quotation, but they wait for the payment to confirm the sale order.

Maybe we can add a general configuration option ? So we can activate or not this behavior.

sebastienbeau commented on sale_quotation/views/sale_view.xml:
```diff
@@ -9,19 +9,90 @@
-                    name="quotation_state"
-                    widget="statusbar"
-                    statusbar_visible="draft,waiting_acceptation,accepted"
-                    attrs="{'invisible': [('state', 'in', ('sale', 'done'))]}"
+                    name="use_customer_quotation_workflow"
+                    widget="boolean_toggle"
+                    options="{'autosave': False, 'terminology': 'Use Customer Quotation Workflow'}"
+                    class="oe_inline"
                 />
```
Following comment is just an opinion so nothing blocking

I do not like a lot to add extra status bar, on previous version I hide the original one and try to be more integrated with existing odoo action.

I fear that this make the UI harder to undertand (many button and status bar)

<img width="1849" height="472" alt="Screenshot 2025-09-04 at 12-34-58 Odoo - S00021" src="https://github.com/user-attachments/assets/1488b439-99c7-4122-b9fc-1131d81d6d93" />

I will show it to @thibaultrey and @dora-jurcevic to get more feedback

nicolas-delbovier-acsone replied:
For the record: we've discuss this in OCA Spain 2026 and **our** clients do like very much this second status bar. It was agreed to keep it as it is for now.

sebastienbeau reviewed (changes requested):
Thanks for the refactor ! Quotation flow will be more powerfull !

nicolas-delbovier-acsone committed:
[FIX] shopinvader_api_quotation: Ensure salesperson is auto-subscribed upon creation

When a quotation is created via the API client, the assigned salesperson (user_id)
was not automatically added as a follower.

This occurred because the Odoo core mechanism for auto-subscribing the user_id
is only triggered when the 'user_id' value is explicitly present in the create
values dictionary. Since the API relies on Odoo's default value computation
to assign the salesperson, the auto-subscription logic was incorrectly bypassed.

This commit explicitly calls message_subscribe for the assigned user_id.partner_id
during the API creation flow, ensuring the salesperson receives relevant Chatter
notifications as expected.

nicolas-delbovier-acsone committed:
[FIX] shopinvader_api_quotation: Ensure salesperson is auto-subscribed upon creation

When a quotation is created via the API client, the assigned salesperson (user_id)
was not automatically added as a follower.

This occurred because the Odoo core mechanism for auto-subscribing the user_id
is only triggered when the 'user_id' value is explicitly present in the create
values dictionary. Since the API relies on Odoo's default value computation
to assign the salesperson, the auto-subscription logic was incorrectly bypassed.

This commit explicitly calls message_subscribe for the assigned user_id.partner_id
during the API creation flow, ensuring the salesperson receives relevant Chatter
notifications as expected.

paradoxxxzero commented on sale_quotation/models/sale_order.py:
```diff
@@ -42,29 +84,166 @@ def _compute_quotation_state(self):
+            if (
+                customer_quotations
+                and self.env.context.get("use_quotation_confirm_wizard")
+                and any(rec.quotation_state != "waiting_acceptation" for rec in self)
```
```suggestion
                and any(rec.quotation_state != "waiting_acceptation" for rec in customer_quotations)
```

paradoxxxzero committed:
sale_quotation: Avoid unnecessary ping pong calls

nicolas-delbovier-acsone committed:
pre-commit update

nicolas-delbovier-acsone committed:
[REF] sale_quotation: refactor confirmation workflow to use native UI prompts

nicolas-delbovier-acsone committed:
[FIX] shopinvader_api_quotation: mssing import in unit tests

nicolas-delbovier-acsone commented:
Can anyone allow the tests to run on this PR?

nicolas-delbovier-acsone committed:
[FIX] shopinvader_api_quotation: adapt client requests to Starlette HTTP test standards

Fix a series of 422 (Unprocessable Entity) test failures caused by an upstream
upgrade to Starlette/FastAPI.

nicolas-delbovier-acsone committed:
[FIX] shopinvader_api_quotation: elevate privileges on quotation acceptance

Fix an AccessError on `procurement.group` when a storefront user accepts
a quotation and `sale_stock` is installed.

paradoxxxzero commented on shopinvader_api_quotation/dependencies.py:
```diff
@@ -0,0 +1,293 @@
+        # triggered by sale_stock can bypass portal/public user ACL limits.
+        return quotation.sudo().action_customer_accept_quotation()
+
+    def _proceess_request_quotation(self, quotation):
```
Typo: should be `_process_request_quotation`

nicolas-delbovier-acsone replied:
Thanks, I'll fix this

paradoxxxzero reviewed (approved):
Minor typo, otherwise LGTM

nicolas-delbovier-acsone committed:
[FIX] shopinvader_api_quotation: fix typo
